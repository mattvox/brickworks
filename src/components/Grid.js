import styled, { css } from 'styled-components';

import media from '../utils/mediaTemplate';
import applyCssWithUnit from '../utils/applyCssWithUnit';
import { gridTypes, baseTypes } from './types';
import { baseStyles } from './Base';

// prettier-ignore
export const gridStyles = css`
  display: flex;
  flex-flow: row wrap;
  justify-content: center;

  ${({
    theme,
    spacing,
    columns,
    xs,
    sm,
    md,
    lg,
    xl,
    breakpoints,
    flow,
    justify,
    alignContent,
    alignItems,
    noSpacing,
  }) => css`
    ${(theme.flow || flow) && css`flex-flow: ${theme.flow || flow};`}
    ${(theme.justify || justify) && css`justify-content: ${theme.justify || justify};`}
    ${(theme.alignContent || alignContent) && css`align-content: ${theme.alignContent || alignContent};`}
    ${(theme.alignItems || alignItems) && css`align-items: ${theme.alignItems || alignItems};`}
    
    ${(columns || xs) && css`
      > .brckwrx-col {flex-basis: ${100 / (columns || xs)}%;}
    `}

    ${sm && media(theme.breakpoints || breakpoints).sm`
      > .brckwrx-col {flex-basis: ${100 / sm}%;}
    `}

    ${md && media(theme.breakpoints || breakpoints).md`
      > .brckwrx-col {flex-basis: ${100 / md}%;}
    `}

    ${lg && media(theme.breakpoints || breakpoints).lg`
      > .brckwrx-col {flex-basis: ${100 / lg}%;}
    `}
    
    ${xl && media(theme.breakpoints || breakpoints).xl`
      > .brckwrx-col {flex-basis: ${100 / xl}%;}
    `}

    ${(theme.noSpacing || noSpacing) || css`
      :not(.brckwrx-row) :not(.brckwrx-col) {
        padding-top: calc(${applyCssWithUnit(spacing || theme.spacing, 'em') || '1em'} / 2);
        padding-bottom: calc(${applyCssWithUnit(spacing || theme.spacing, 'em') || '1em'} / 2);
        
        .brckwrx-row {
          padding-left: calc(${applyCssWithUnit(spacing || theme.spacing, 'em') || '1em'} / 2);
          padding-right: calc(${applyCssWithUnit(spacing || theme.spacing, 'em') || '1em'} / 2);
        }

        .brckwrx-col {
          padding: calc(${applyCssWithUnit(spacing || theme.spacing, 'em') || '1em'} / 2);
        }
      }
    `}
  `};
`;

// prettier-ignore
const Grid = styled.div.attrs({ className: `brckwrx-grid` })`
  ${baseStyles}
  ${gridStyles}
`;

Grid.propTypes = {
  ...gridTypes,
  ...baseTypes,
};

Grid.nav = Grid.withComponent('nav');
Grid.aside = Grid.withComponent('aside');
Grid.main = Grid.withComponent('main');
Grid.header = Grid.withComponent('header');
Grid.footer = Grid.withComponent('footer');
Grid.section = Grid.withComponent('section');

export default Grid;
