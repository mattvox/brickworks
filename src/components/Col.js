import styled, { css } from 'styled-components';

import applyFlex from '../utils/applyFlex';
import { colTypes, baseTypes, gridTypes } from './types';
import { gridStyles } from './Grid';
import { baseStyles } from './Base';

// prettier-ignore
const Col = styled.div.attrs({ className: `brckwrx-col` })`
  margin: 0;
  ${baseStyles}
  flex: ${({ flex }) => applyFlex(flex) || '1 1 auto'};
  ${({ order, alignSelf, grid }) => css`
    ${order && css`order: ${order};`}
    ${alignSelf && css`align-self: ${alignSelf};`}
    ${grid && gridStyles}
    ${grid && css`padding: 0 !important;`}
    .brckwrx-row {
      padding-left: 0 !important;
      padding-right: 0 !important;
    }
  `}
`;

Col.propTypes = {
  ...baseTypes,
  ...colTypes,
  ...gridTypes,
};

Col.nav = Col.withComponent('nav');
Col.aside = Col.withComponent('aside');
Col.main = Col.withComponent('main');
Col.header = Col.withComponent('header');
Col.footer = Col.withComponent('footer');
Col.section = Col.withComponent('section');

export default Col;
