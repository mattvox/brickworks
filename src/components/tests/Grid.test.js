import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { css } from 'styled-components';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import testStyle from './helpers/testStyle';
import Grid from '../Grid';
import Col from '../Col';

Enzyme.configure({ adapter: new Adapter() });

describe('<Grid />', () => {
  const wrapper = shallow(<Grid />);

  it('should render correctly', () => {
    expect(wrapper.find('div').exists()).toBe(true);
  });

  describe('with default props/settings', () => {
    const expectedStyleList = [
      { style: 'display', value: 'flex' },
      { style: 'flex-flow', value: 'row wrap' },
      { style: 'justify-content', value: 'center' },
    ];

    it('should match snapshot', () => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    expectedStyleList.forEach(({ style, value }) => {
      return testStyle(wrapper, style, value);
    });
  });

  describe('with added props', () => {
    const wrapperWithProps = shallow(
      <Grid
        flow="column nowrap"
        justify="center"
        alignContent="space-around"
        alignItems="flex-end"
        badProp1="this is a bad prop"
        badProp2="this is another bad prop"
        spacing
      />
    );

    const expectedStyleList = [
      { style: 'flex-flow', value: 'column nowrap' },
      { style: 'justify-content', value: 'center' },
      { style: 'align-content', value: 'space-around' },
      { style: 'align-items', value: 'flex-end' },
    ];

    it('should match snapshot with props', () => {
      expect(shallowToJson(wrapperWithProps)).toMatchSnapshot();
    });

    expectedStyleList.forEach(({ style, value }) => {
      return testStyle(wrapperWithProps, style, value);
    });
  });

  describe('passes styles down to children columns', () => {
    const wrapper = mount(
      <Grid spacing={2}>
        <Col>Hello</Col>
      </Grid>
    );

    Col.displayName = 'Col';

    it('should match snapshot with props', () => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it(`should pass padding when spacing prop is present`, () => {
      const tree = renderer
        .create(
          <Grid spacing={2}>
            <Col />
          </Grid>
        )
        .toJSON();

      // prettier-ignore
      expect(tree).toHaveStyleRule('padding', 'calc(2em / 2)', {
        modifier: css`:not(.brckwrx-row) :not(.brckwrx-col) .brckwrx-col`,
      });
    });

    it(`should pass flex when the columns prop is present`, () => {
      const tree = renderer
        .create(
          <Grid columns={2}>
            <Col />
          </Grid>
        )
        .toJSON();

      // prettier-ignore
      expect(tree).toHaveStyleRule('flex-basis', '50%', {
        modifier: css`> .brckwrx-col`,
      });
    });

    it(`should pass flex and adjust flex-basis based on media queries when passed xs, sm, md, lg, and xl props`, () => {
      const tree = renderer
        .create(
          <Grid xs={4} sm={1} md={2} lg={4} xl={10}>
            <Col />
          </Grid>
        )
        .toJSON();

      // prettier-ignore
      expect(tree).toHaveStyleRule('flex-basis', '25%', {
        modifier: css`> .brckwrx-col`,
      });

      // prettier-ignore
      expect(tree).toHaveStyleRule('flex-basis', '100%', {
        media: 'screen and (min-width: 576px)',
        modifier: css`> .brckwrx-col`,
      });

      // prettier-ignore
      expect(tree).toHaveStyleRule('flex-basis', '50%', {
        media: 'screen and (min-width: 768px)',
        modifier: css`> .brckwrx-col`,
      });

      // prettier-ignore
      expect(tree).toHaveStyleRule('flex-basis', '25%', {
        media: 'screen and (min-width: 992px)',
        modifier: css`> .brckwrx-col`,
      });

      // prettier-ignore
      expect(tree).toHaveStyleRule('flex-basis', '10%', {
        media: 'screen and (min-width: 1200px)',
        modifier: css`> .brckwrx-col`,
      });
    });
  });
});
