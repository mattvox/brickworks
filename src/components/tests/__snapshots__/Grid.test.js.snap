// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> passes styles down to children columns should match snapshot with props 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0:not(.brckwrx-row) :not(.brckwrx-col) {
  padding-top: calc(2em / 2);
  padding-bottom: calc(2em / 2);
}

.c0:not(.brckwrx-row) :not(.brckwrx-col) .brckwrx-row {
  padding-left: calc(2em / 2);
  padding-right: calc(2em / 2);
}

.c0:not(.brckwrx-row) :not(.brckwrx-col) .brckwrx-col {
  padding: calc(2em / 2);
}

.c1 {
  margin: 0;
  box-sizing: border-box;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

.c1 .brckwrx-row {
  padding-left: 0 !important;
  padding-right: 0 !important;
}

<styled.div
  spacing={2}
>
  <div
    className="brckwrx-grid c0"
    spacing={2}
  >
    <Col>
      <div
        className="brckwrx-col c1"
      >
        Hello
      </div>
    </Col>
  </div>
</styled.div>
`;

exports[`<Grid /> with added props should match snapshot with props 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-content: space-around;
  -ms-flex-line-pack: space-around;
  align-content: space-around;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c0:not(.brckwrx-row) :not(.brckwrx-col) {
  padding-top: calc(1em / 2);
  padding-bottom: calc(1em / 2);
}

.c0:not(.brckwrx-row) :not(.brckwrx-col) .brckwrx-row {
  padding-left: calc(1em / 2);
  padding-right: calc(1em / 2);
}

.c0:not(.brckwrx-row) :not(.brckwrx-col) .brckwrx-col {
  padding: calc(1em / 2);
}

<div
  className="brckwrx-grid c0"
  spacing={true}
/>
`;

exports[`<Grid /> with default props/settings should match snapshot 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0:not(.brckwrx-row) :not(.brckwrx-col) {
  padding-top: calc(1em / 2);
  padding-bottom: calc(1em / 2);
}

.c0:not(.brckwrx-row) :not(.brckwrx-col) .brckwrx-row {
  padding-left: calc(1em / 2);
  padding-right: calc(1em / 2);
}

.c0:not(.brckwrx-row) :not(.brckwrx-col) .brckwrx-col {
  padding: calc(1em / 2);
}

<div
  className="brckwrx-grid c0"
/>
`;
